accel/kvm/vm-introspection.c

```diff
@@ -124,46 +144,122 @@ static bool do_handshake(CharBackend *sock, Object *key, const char *sock_name,
     return true;
 }
 
-int vm_introspection_fd(Object *obj, uint32_t *commands, uint32_t *events,
-                        Error **errp)
+static bool connect_fd(VMIntrospection *i, Error **errp)
 {
-    VMIntrospection *i = VM_INTROSPECTION(obj);
-    Chardev *chr;
-    CharBackend sock;
-    Object *key;
-    int fd = -1;
-
+    memset(&i->h, 0, sizeof(i->h));
+    i->h.fd = -1;
     /* TODO: proper handling of allow,deny props */
-    *commands = *events = -1;
+    i->h.commands = i->h.events = -1;
 
-    key = object_resolve_path_component(object_get_objects_root(), i->keyid);
-    if (!key) {
-        error_setg(errp, "No secret object with id '%s'", i->keyid);
-        return -1;
+    if (do_handshake(&i->sock, i->key, i->chardevid, errp)) {
+        i->h.fd = object_property_get_int(OBJECT(i->chr), "fd", errp);
+
+        if (i->h.fd == -1) {
+            error_append_hint(errp, "no file handle from '%s'", i->chardevid);
+        }
     }
 
-    chr = qemu_chr_find(i->chardevid);
-    if (chr == NULL) {
-        error_setg(errp, "Device '%s' not found", i->chardevid);
-        return -1;
+    return (i->h.fd != -1);
+}
+
+static bool connect_introspection(VMIntrospection *i, Error **errp)
+{
+    int ret;
+
+    if (!connect_fd(i, errp)) {
+        error_append_hint(errp, "introspection handshake failed\n");
+        return false;
     }
 
-    if (!qemu_chr_fe_init(&sock, chr, &error_abort)) {
-        error_setg(errp, "Device '%s' not initialized", i->chardevid);
-        return -1;
+    ret = kvm_vm_ioctl(i->kvm, KVM_INTROSPECTION, &i->h);
+
+    if (ret < 0) {
+        error_setg_errno(errp, -errno, "ioctl/KVM_INTROSPECTION failed");
+        return false;
     }
 
-    if (do_handshake(&sock, key, i->chardevid, errp)) {
-        fd = object_property_get_int(OBJECT(chr), "fd", errp);
-        if (fd != -1) {
-            fd = dup(fd);
-        } else {
-            error_setg(errp, "no file handle from '%s': %d", i->chardevid,
-                       errno);
+    return true;
+}
+
+static void *reconnect_introspection(void *arg)
+{
+    VMIntrospection *i = arg;
+    int ret = -1;
+
+    for (; !i->stopping && ret < 0; sleep(2)) {
+        Error *err = NULL;
+
+        ret = qemu_chr_fe_wait_connected(&i->sock, &err);
+
+        if (ret == 0 && connect_introspection(i, &err)) {
+            info_report("introspection connected");
+            break;
+        }
+
+        if (err) {
+            warn_report_err(err);
         }
+        /* error_free(err); */
```

Comment:
Left commented

Reply:
I'm thinking of dropping the error processing from that function, because these errors:
 - can't be seen with the -demonize option
 - are noisy without -demonize
 - are somehow useless
I should ignore any error and just retry.

Reply:
Errors the User can tackle should always be reported (example: wrong file permission, wrong configuration). If there are no such case then we can ignore them.

For how long is going to retry? Eventually the User would like to know why the app is hanging so he/she can disable VMI in case is offending.

Reply:
If the user doesn't add ",reconnect=N" to the chardev/socket arguments, the error message will come from QEMU (that it can't create the device) and the VM won't be started.

If the user adds ",reconnect=N", the introspection will try to connect when started, and if it fails, it will show the error message "introspection socket is not open\n it will reconnect when ready\n". It will retry until successful, and it will inform when the connection take place. I'm not sure how helpful are the retry error messages.

Reply:
If the retry logic doesn't take long, I'd leave the messages out. In case the retry logic might take long it's usually a good practice to inform the User to prevent him/her from giving up.

A good middle-ground is printing a message every X attempts and, in case of too many attempt, reporting the last error message (example: "Connection reset by peer").

Reply:
I agree with you. But, in this case, when half of that function deals with error messages and the use cases that will benefit from them are not clear enough, I would leave it as it is for now. Not to mention that the user has another feedback from the introspector when it fails to connect.

Maybe adding a property to the introspection object that will hold the connection status/error it will be more helpful. This property could be queried (qmp?) and the introspector could show more info when it fails to connect (if it has already a connection to QEMU). Or maybe not. Let's see the real use cases that will benefit from these error messages and we will know the best way to handle them.